### ISSUE TITLE
Research about queueing code using PHP

### ISSUE THREAD

**Issue by catherineunciano** [IT3A-2016/Paper-Icebergs#18]

**Comment by catherineunciano** [comment]
These links will be very helpful:
https://www.codementor.io/smorhaim/creating-simple-powerful-queue-system-beanstalkd-microservices-du107kllb
http://www.endyourif.com/building-a-scalable-queueing-system-with-php/
https://github.com/geberele/php-queue-system
https://github.com/CoderKungfu/php-queue

**Comment by catherineunciano** [comment]
First things first. We need what I call a publisher and then a number of consumers / workers.

Get this library into your composer:

"pda/pheanstalk": "~3.0.2",
I created a simple class called Queue\Publisher.php and Queue\Consumer.php:

namespace Queue;

use Pheanstalk\Pheanstalk;

class Publisher
{
    private $queue;

    public function __construct(array $args)
    {
        $this->config = ['queue' => ['host' => 'yourdomain.com']; // Don't mind this. I typically have a config file, but I moved it here for purposes of this tutorial. Use IP or Domain for the host. 

        $this->queue = $args['queue']; // Just a name for the queue. I pass this as a parameter. Beanstalkd calls the queues "tubes" but this really is just how you want to call it. If it doesn't exist, it gets created. 
        $this->client = new Pheanstalk($this->config['queue']['host']); // Instantiate an object. 

    }

    public function send($request)
    {
        return $this->client
            ->useTube($this->queue)
            ->put(json_encode($request)); // Send anything you want, json encoded and that's it!
    }
}
And for the consumer:

namespace Queue;

use Pheanstalk\Pheanstalk;

class Consumer
{

    private $queue;

    public function __construct(array $args)
    {
        $this->config = ['queue' => ['host' => 'yourdomain.com']; // Don't mind this. I typically have a config file, but I moved it here for purposes of this tutorial. Use IP or Domain for the host. 

        $this->queue = $args['queue'];
        $this->client = new Pheanstalk($this->config['queue']['host']);
    }

    public function listen()
    {
        $this->client->watch($this->queue); // Pass the name of the queue again. 

        while ($job = $this->client->reserve()) { // Do this forever... so it's always listening. 
            $message = json_decode($job->getData(), true); // Decode the message 

            $status = $this->process($message);

            if($status)
                $this->client->delete($job);
            else
                $this->client->delete($job); // Delete anyway. You could burry it, meaning it gets re-tried later. 

        }
    }

    public function process($msg)
    {
         // Do some operation and return true if success or false
    }
Finally of course you need to call this:

$publish = new Queues\Publisher(['queue' => "someQueue"]);
$publish->send("Do this..."); // This could be an array, a class name, etc.
Then in the other side, retrieve it

$consumer = new Queues\Consumer(['queue' => "someQueue"]); // Basically this will run via console usually.
And that's it! In our application we run PhalconPHP and we have the concept of Tasks, we use these tasks tu run the Consumer while we use a Cronjob to "publish" messages. In the same way, we can run a large number of consumer tasks to process the queue faster if we need to.

For even more scalability, we could setup any number of servers with the same code, point the consumer to the same Beanstalkd IP/Domain and run even more consumers so we have virtually unlimited space to grow.

As a tip, we are running our tasks via Supervisor, so for example we have 10 workers running which clears our queues in about 15 minutes. (The delay takes place in the external API's we query for each of the thousand of messages we process). If your operations are local or you can control the performance of them, you could be processing hundreds of messages per second as needed.

**Comment by catherineunciano** [comment]
Building A Scalable Queueing System With PHP

Imagine an admin area of a website that allows the administrator to send out a mass email to all of their users.  The simple process to building this functionality would be as follows:

1. Build a form that accepts a subject and a body for the email.
2. Retrieve the list of users from your database.
3. Loop through the users and send each person an individual email.

The above example works nice and fast when there are only a few hundred users.  However, imagine trying to send this email to 10,000 users.  The administrator would be waiting a long time for this process to finish.  Not only that, if they closed the browser, it probably would not finish properly.

So, the goal of our queueing system is to remove a specific process from running "online" (in a web browser) and running it "offline" with a scheduled task.

Setup

The goal of this article is to provide a theoritical example of a queueing system and allowing that system to accommodate large scale processing.  To implement our above example, I am going to assume that you have an existing web server and web site already running.  We will also need a process similar to the above that you wish to perform the offline queue process.

By the end of this article, you should be able to (with minimal effort), convert your existing process into an efficiently scaled queueing process.

Other great examples of a queueing system are:

- third party tracking data
- RSS content creation
- RSS content reading
- etc...
Implementation

To begin, we need to create a database table.  This table will be used to maintain our queue and allow for a first in, first out process.  Depending on the queue you are building, your table might be slightly different.  Sticking with our example, our table should looking something similar to the following:
queue:

- id int AUTO INCREMENT PRIMARY KEY
- to varchar(255) NOT NULL
- from varchar(255) NOT NULL
- subject varchar(255) NOT NULL
- body TEXT NOT NULL
- created DATETIME NOT NULL
- sent DATETIME NULL

Each time the admin user wishes to send an email to their users, we will insert one row into our queue table PER USER.  The row will contain the subject, body, created date, as well as the to and from.  Our id will be auto generated and most importantly, our sent date will be null (more on this later).

Now that our table is created, begin by updating your current process to no longer perform the email.  Copy and paste this code into a separate file for later as we'll need it when we implement our queue.  Replace that process with a SQL query to insert into our queue table instead (the loop through all users will still be required), e.g.:

<?php
// assuming we already have an existing database connection...
// after our form post that collects subject and body...
$result = mysql_query('SELECT email FROM users');
// loop through all users
while ($row = mysql_fetch_assoc($result)) {
// insert one row in the queue per user
mysql_query('INSERT INTO queue (
null,
\'' . $row['email'] . '\',
\'noreply@mydomain.com\',
\'' . $_POST['subject'] . '\',
\'' . $_POST['body'] . '\',
\'' . Date('Y-m-d H:i:s') . '\',
null
)');
}
?>


Our queue table will now begin to fill up when an admin user wishes to send out emails.  Now we need to create our queue processing page.  Let's create a new file called queue.php:

<?php
// ensure the file doesn't stop processing
set_time_limit(0);
// assuming we already have an existing database connection...
// get all of our pending emails to be sent
$result = mysql_query('SELECT * FROM queue WHERE sent IS NULL');
// loop through each one
while ($row = mysql_fetch_assoc($result)) {
// put existing process to send email here...
// after processing email, mark it as sent
mysql_query('UPDATE queue SET sent = \'' . Date('Y-m-d H:i:s') . '\' WHERE id = ' . $row['id']);
}
?>


The above code is just a basic example of queue processing.  It begins by retrieving a list of outstanding queues and processes them one-by-one.  After each queue is finished, we set the sent date to "now" to avoid re-processing this queue later.  We'll also need to copy and paste our previous code that we removed a few steps ago to perform the actual email sending.

To finish our queueing process, we need to create a scheduled task to run this process regularly.  The easiest way to run this script is using PHP in command line mode (why involve Apache when we don't need too).  If we are on Windows, it would be something similar to (in a command prompt):

C:\xammp\php\php.exe C:\xammp\htdocs\queue.php.

Basically, we are invoking PHP and telling it to run our script.  In the example above be sure to edit both locations to your files as necessary.

Create the scheduled task now and let's set it to run once per day (or any other interval that is appropriate for your queueing needs).  For help creating a scheduled task (or cron job), do a quick Google search.

That's it, our queueing process is now completed.  Each time it runs, it will retrieve a list of pending emails to send out and update the sent time after it is done preventing it from being processed again.
Scaling

Let's shift our focus to scaling our queueing system.  Our current limitation is how many emails our web server can send out per minute.  Imagine we need to send out 100,000 emails and our server can send out approximately 500 emails per minute.  It would take roughly 3+ hours to complete this process.  Now imagine 1,000,000 emails.  We are now talking 30+ hours to complete!  That would prevent sending daily emails because it would take more than one day to complete the entire queue.

We can solve this problem easily by scaling our web servers and adding more into the mix.  This is a very typical solution, add more web servers and evenly distribute the load to each server to not overload any one server.

In our above example, adding one server should cut the time down in half, which would allow for a daily email to 1,000,000 users to occur.  So we'll assume we now have two web servers that will be used to process our queue.

The next step is to ensure the two servers "don't step on each others toes", e.g. send out the same email twice.  Most users wouldn't enjoy constantly getting two identical emails.  The best way I've found to accomplish this is to alter our original queue table and add the following field:

- server varchar(50) NOT NULL

Now, when we create our email queue, we can evenly distribute the emails across the multiple web servers.  With this new field we will need to update our process that inserts into the database as follows:

// assuming we already have an existing database connection...
// define our array of servers (replace with names of your servers)
$server = array('WEB1', 'WEB2');
// after our form post that collects subject and body...
$result = mysql_query('SELECT email FROM users');
// loop through all users
while ($row = mysql_fetch_assoc($result)) {
// pick a random number between 0 and count of servers - 1
$rand = mt_rand(0, count($server) - 1);
// insert one row in the queue per user
mysql_query('INSERT INTO queue (
null,
\'' . $row['email'] . '\',
\'noreply@mydomain.com\',
\'' . $_POST['subject'] . '\',
\'' . $_POST['body'] . '\',
\'' . Date('Y-m-d H:i:s') . '\',
null,
\'' . $server[$rand] . '\'
)');
}
?>


The above code simply picks a random server to assign the queue to.  This selection is saved into the "server" column.  Over a large number of users our random generated number should maintain an even balance across all servers.

Now we need to update our queue processing file as follows:

<?php
// ensure the file doesn't stop processing
set_time_limit(0);
// assuming we already have an existing database connection...
// get all of our pending emails to be sent for THIS hostname
$result = mysql_query('SELECT * FROM queue WHERE sent IS NULL AND server = \'' . `hostname` . '\'');
// loop through each one
while ($row = mysql_fetch_assoc($result)) {
// put existing process to send email here...
// after processing email, mark it as sent
mysql_query('UPDATE queue SET sent = \'' . Date('Y-m-d H:i:s') . '\' WHERE id = ' . $row['id']);
}
?>


The above change is a little more subtle, can you spot it?  We've updated the process to get our pending emails to only get ones assigned to our current server.  `hostname` returns us the name of our web server specified in our Apache configuration.  Once this change is done, create the scheduled task on the new web server(s) to run at the same time and interval as the previous scheduled task.

As your user base grows and more processing is required, we now just need to add more web servers to the loop and our code will take care of the rest!
Summary

In today's article we discussed how to convert an "online" process that is slowing our web site down and convert it into an "offline" queueing system.  Not only did we do that, we implemented an extremely simple solution to grow our queueing system to accommodate a large scale queue.

**Comment by catherineunciano** [comment]
An example of a Queue System in PHP

Optimise the performances of a server for a better user experience is always an important factor. Sometimes users have to wait for an answer from the server because of long processes that are happening all at once in the back end. In order to avoid this is important to consider that not all of them have to be executed at the same time before the server replies to each request. Certainly we can postpone some of them, especially the ones that have job tasks with longer loading time, and put them on hold into a processes queue list that will be executed later on.

queue list

The Queue List can be executed whenever we think it's better for the web site. For example it could be at night if the tasks require long time and we'd like to affect as less users as possible in case the website performances slow down.

The structure of a workflow that represents the execution of a queue list looks like the following:

flowchart

Translating in PHP code the above flowchart we'll get something like this:

<?php

/**
 * Example of a Queue System.
 */
function queue_system_example() {
  require_once(dirname(__FILE__) . '/lib/QueueClass.php');
  $queue = new QueueClass();
  // Populate the queue.
  for ($i = 0; $i < 100; $i++) {
    $data = range(0, $i + 1);
    $queue->createItem($data);
  }
  $jobs_to_do = TRUE;
  $start = microtime(true);

  try {
    while ($jobs_to_do) {
      $item = $queue->claimItem();

      if ($item) {
        echo 'Processing the item ' . $item->item_id . '...' . PHP_EOL;

        // Execute the job task in a different function.
        if (execute_the_job_task($item)) {
          // Delete the item.
          $queue->deleteItem($item);
          echo 'Item ' . $item->item_id . ' processed.' . PHP_EOL;
        }
        else {
          // Release the item to execute the job task again later.
          $queue->releaseItem($item);
          echo 'Item ' . $item->item_id . ' NOT processed.' . PHP_EOL;
          $jobs_to_do = FALSE;
          echo 'Queue not completed. Job task not executed.' . PHP_EOL;
        }
      }
      else {
        $jobs_to_do = FALSE;
        $time_elapsed = microtime(true) - $start;
        $number_of_items = $queue->numberOfItems();
        if ($number_of_items == 0) {
          echo 'Queue completed in ' . $time_elapsed . ' seconds.' . PHP_EOL;
        }
        else {
          echo 'Queue not completed, there are ' . $number_of_items . ' items left.' . PHP_EOL;
        }
      }
    }
  }
  catch (Exception $e) {
    echo $e->getMessage();
  }
}

/**
 * Execute the job task.
 */
function execute_the_job_task($item) {
  // Do something with the item.
  return TRUE;
}

For this example the QueueClass created and used is:

<?php

/**
 * @file
 * QueueClass Class.
 */

/**
 * Static queue implementation.
 */
class QueueClass {
  /**
   * The queue data.
   *
   * @var array
   */
  protected $queue;

  /**
   * Counter for item ids.
   *
   * @var int
   */
  protected $id_sequence;

  /**
   * Start working with a queue.
   */
  public function __construct() {
    $this->queue = array();
    $this->id_sequence = 0;
  }

  /**
   * Add a queue item and store it directly to the queue.
   */
  public function createItem($data) {
    $item = new stdClass();
    $item->item_id = $this->id_sequence++;
    $item->data = $data;
    $item->created = time();
    $item->expire = 0;
    $this->queue[$item->item_id] = $item;
  }

  /**
   * Retrieve the number of items in the queue.
   */
  public function numberOfItems() {
    return count($this->queue);
  }

  /**
   * Claim an item of the queue for a specific time.
   */
  public function claimItem($lease_time = 3600) {
    foreach ($this->queue as $key => $item) {
      if ($item->expire == 0) {
        $item->expire = time() + $lease_time;
        $this->queue[$key] = $item;
        return $item;
      }
    }
    return FALSE;
  }

  /**
   * Delete a finished item from the queue.
   */
  public function deleteItem($item) {
    unset($this->queue[$item->item_id]);
  }

  /**
   * Release an item that the worker could not process, so another
   * worker can come in and process it before the timeout expires.
   */
  public function releaseItem($item) {
    if (isset($this->queue[$item->item_id]) && $this->queue[$item->item_id]->expire != 0) {
      $this->queue[$item->item_id]->expire = 0;
      return TRUE;
    }
    return FALSE;
  }

  /**
   * Create a queue.
   */
  public function createQueue() {
    // Nothing needed here.
  }

  /**
   * Delete a queue.
   */
  public function deleteQueue() {
    $this->queue = array();
    $this->id_sequence = 0;
  }

}
This queue class (that implements the interface DrupalQueueInterface) is useful for static implementation, for a dynamic implementation instead is necessary to store the queue list in a database.

**Comment by catherineunciano** [comment]
PHP-Queue

Gitter

A unified front-end for different queuing backends. Includes a REST server, CLI interface and daemon runners.

Build Status

Why PHP-Queue?

The pains of implementing a queueing system (eg. Beanstalk, Amazon SQS, RabbitMQ) for your application:

Which one is most efficient? Performant?
Learning curve to effectively implement the queue backend & the libraries.
Time taken to develop the application codes.
Vendor locked in, making it impossible to switch.
Requires massive code change (ie. not flexible) when use case for the queue changes.
PHP-Queue hopes to serve as an abstract layer between your application code and the implementation of the queue.

Benefits

Job Queue is Backend agnostic

Just refer to the queue by name, what it runs on is independent of the application code. Your code just asks for the next item in the PHPQueue\JobQueue, and you'll get a PHPQueue\Job object with the data and jobId.

Flexible Job Queue implementation

You can decide whether each PHPQueue\JobQueue only carries 1 type of work or multiple types of target workers. You control the retrieval of the job data from the Queue Backend and how it is instantiated as a PHPQueue\Job object. Each PHPQueue\Job object carries information on which workers it is targeted for.

Independent Workers

Workers are independent of Job Queues. All it needs to worry about is processing the input data and return the resulting data. The queue despatcher will handle the rest. Workers will also be chainable.

Powerful

The framework is deliberately open-ended and can be adapted to your implementation. It doesn't get in the way of your queue system.

We've build a simple REST server to let you post job data to your queue easily. We also included a CLI interface for adding and triggering workers. All of which you can sub-class and overwrite.

You can also include our core library files into your application and do some powerful heavy lifting.

Several backend drivers are bundled:

Memcache
Redis
MongoDB
CSV These can be used as the primary job queue server, or for abstract FIFO or key-value data access.
Installation

Installing via Composer

Composer is a dependency management tool for PHP that allows you to declare the dependencies your project needs and installs them into your project. In order to use the PHP-Queue through Composer, you must do the following:

Add "coderkungfu/php-queue" as a dependency in your project's composer.json file. Visit the Packagist page for more details.

Download and install Composer.

curl -s "http://getcomposer.org/installer" | php
Install your dependencies.

php composer.phar install
All the dependencies should be downloaded into a vendor folder.

Require Composer's autoloader.

<?php
require_once '/path/to/vendor/autoload.php';
?>
Getting Started

You can have a look at the Demo App inside .\vendor\coderkungfu\php-queue\src\demo\ folder for a recommended folder structure.

htdocs folder
.htaccess
index.php
queues folder
<QueueNameInCamelCase>Queue.php
workers folder
<WorkerNameInCamelCase>Worker.php
runners folder
cli.php file
config.php file
I would also recommend putting the autoloader statement and your app configs inside a separate config.php file.

Recommended config.php file content:

<?php
require_once '/path/to/vendor/autoload.php';
PHPQueue\Base::$queue_path = __DIR__ . '/queues/';
PHPQueue\Base::$worker_path = __DIR__ . '/workers/';
?>
Altenative config.php file:

You can also declare your application's namespace for loading the Queues and Workers.

<?php
require_once '/path/to/vendor/autoload.php';
PHPQueue\Base::$queue_namespace = '\MyFabulousApp\Queues';
PHPQueue\Base::$worker_namespace = '\MyFabulousApp\Workers';
?>
PHP-Queue will attempt to instantiate the PHPQueue\JobQueue and PHPQueue\Worker classes using your namespace - appended with the queue/worker name. (ie. \MyFabulousApp\Queues\Facebook).

It might be advisable to use Composer's Custom Autoloader for this.

Note:
If you declared PHPQueue\Base::$queue_path and/or PHPQueue\Base::$worker_path together with the namespace, the files will be loaded with require_once from those folder path AND instantiated with the namespaced class names.

REST Server

The default REST server can be used to interface directly with the queues and workers.

Copy the htdocs folder in the Demo App into your installation. The index.php calls the \PHPQueue\REST::defaultRoutes() method - which prepares an instance of the Respect\Rest REST server. You might need to modify the path of config.php within the index.php file.

Recomended installation: use a new virtual host and map the htdocs as the webroot.

Add new job.

# Form post
curl -XPOST http://localhost/<QueueName>/ -d "var1=foo&var2=bar"
# JSON post
curl -XPOST http://localhost/<QueueName>/ -H "Content-Type: application/json" -d '{"var1":"foo","var2":"bar"}'
Trigger next job.

curl -XPUT http://localhost/<QueueName>/
Read the full documentation on Respect\Rest to further customize to your application needs (eg. Basic Auth).

Command Line Interface (CLI)

Copy the cli.php file from the Demo App into your installation. This file implements the \PHPQueue\Cli class. You might need to modify the path of config.php within the cli.php file.

Add new job.

$ php cli.php <QueueName> add --data '{"boo":"bar","foo":"car"}'
Trigger next job.

$ php cli.php <QueueName> work
You can extend the PHPQueue\Cli class to customize your own CLI batch jobs (eg. import data from a MySQL DB into a queue).
